core/templates/dev/head/domain/objects/NumberWithUnitsObjectFactory.js

```
@@ -0,0 +1,255 @@
+});
+
+oppia.factory('NumberWithUnitsObjectFactory', [
+  'UnitsObjectFactory', 'FractionObjectFactory',
+  'NUMBER_WITH_UNITS_PARSING_ERRORS', function(
+      UnitsObjectFactory, FractionObjectFactory,
+      NUMBER_WITH_UNITS_PARSING_ERRORS) {
+    var NumberWithUnits = function(type, real, fractionObj, unitsObj) {
+      this.type = type;
+      this.real = real;
+      this.fraction = fractionObj;
+      this.units = unitsObj;
+    };
+
+    NumberWithUnits.prototype.toString = function() {
+      var numberWithUnitsString = '';
+      if (this.units.toString().includes('$')) {
+        this.units.units = this.units.toString().replace('$', '');
+        numberWithUnitsString += '$' + ' ';
+      }
+      if (this.units.toString().includes('Rs')) {
+        this.units.units = this.units.toString().replace('Rs', '');
+        numberWithUnitsString += 'Rs' + ' ';
+      }
+
+      if (this.type === 'real') {
+        numberWithUnitsString += this.real + ' ';
+      } else {
+        numberWithUnitsString += this.fraction.toString() + ' ';
+      }
+      numberWithUnitsString += this.units.toString();
+      numberWithUnitsString = numberWithUnitsString.trim();
+
+      return numberWithUnitsString;
+    };
+
+    NumberWithUnits.fromRawInputString = function(rawInput) {
+      rawInput = rawInput.trim();
+      var type = '';
+      var real = '';
+      var fractionObj = '';
+      var units = '';
+      var value = 0;
+      var unitObj = '';
+
+      // Start with digit when there is no currency unit.
+      if (rawInput.match(/^\d/)) {
+        var ind = rawInput.indexOf(rawInput.match(/[a-z(]/i));
+        if (ind === -1) {
+          // There is value with no units.
+          value = rawInput;
+          units = '';
+        } else {
+          value = rawInput.substr(0, ind).trim();
+          units = rawInput.substr(ind).trim();
+        }
+      } else {
+        var ind = rawInput.indexOf(' ');
```

Use a simple regex here to match `/(\$\s|Rs\s|₹\s)/` and find index like that.

---

I'm doing validation on currencies also. In this case, I need to change the whole lot of code. I think what I implemented is also correct as far as I'm validating currency units.

---

I don't think so. Add a test case `kg 2 s^2` where it should raise an error in frontend tests. Also, I don't see how finding index using regex affects a whole bunch of code (you're doing the same thing in IF part above).

---

Done. Now, currency units are validated explicitly. Added the test case too.